control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go

```diff
@@ -96,38 +96,74 @@ func (r *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	// even if Endpoints are empty or have no mesh pod, iff. the service has a selector.
 	// This should ensure that we don't target kube or consul (system) services.
-	if workloadSelector != nil {
-		//TODO: Maybe check service-enable label here on service/deployments/other pod owners
-		if err = r.registerService(ctx, resourceClient, service, workloadSelector); err != nil {
-			// We could be racing with the namespace controller.
-			// Requeue (which includes backoff) to try again.
-			if common.ConsulNamespaceIsNotFound(err) {
-				r.Log.Info("Consul namespace not found; re-queueing request",
-					"service", service.GetName(), "ns", req.Namespace,
-					"consul-ns", r.getConsulNamespace(req.Namespace), "err", err.Error())
-				return ctrl.Result{Requeue: true}, nil
-			}
-			errs = multierror.Append(errs, err)
+	if consulSvc.Workloads == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Register the service in Consul.
+	//TODO: Maybe check service-enable label here on service/deployments/other pod owners
+	if err = r.registerService(ctx, resourceClient, service, consulSvc); err != nil {
+		// We could be racing with the namespace controller.
+		// Requeue (which includes backoff) to try again.
+		if common.ConsulNamespaceIsNotFound(err) {
+			r.Log.Info("Consul namespace not found; re-queueing request",
+				"service", service.GetName(), "ns", req.Namespace,
+				"consul-ns", r.getConsulNamespace(req.Namespace), "err", err.Error())
+			return ctrl.Result{Requeue: true}, nil
 		}
+		return ctrl.Result{}, err
 	}
-	return ctrl.Result{}, errs
+
+	return ctrl.Result{}, nil
 }
 
-// getWorkloadSelectorFromEndpoints calculates a Consul service WorkloadSelector from Endpoints based on pod names and
-// owners.
-func (r *Controller) getWorkloadSelectorFromEndpoints(ctx context.Context, pf PodFetcher, endpoints *corev1.Endpoints) (*pbcatalog.WorkloadSelector, error) {
-	podPrefixes := make(map[string]any)
-	podExactNames := make(map[string]any)
+func (r *Controller) getConsulService(ctx context.Context, pf PodFetcher, service corev1.Service, endpoints corev1.Endpoints) (*pbcatalog.Service, error) {
+	prefixedPods, exactNamePods, err := r.getWorkloadDataFromEndpoints(ctx, pf, endpoints)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create Consul Service resource to be registered.
+	return &pbcatalog.Service{
+		Workloads:  getWorkloadSelector(prefixedPods, exactNamePods),
+		Ports:      getServicePorts(service, prefixedPods, exactNamePods),
+		VirtualIps: r.getServiceVIPs(service),
+	}, nil
+}
+
+type podSetData struct {
+	podCount  int
+	samplePod *corev1.Pod
+}
+
+// selectorPodData represents data for each set of pods represented by a WorkloadSelector value.
+// The data may be for several pods (prefix) or a single pod (exact name).
+// This is used for choosing the ideal Consul service TargetPort value when the K8s service target port is numeric.
+type selectorPodData map[string]*podSetData
```

Comment:
Added a type for this to simplify passing it around and make the code more legible.

Below, we use the `podSetData` struct to accumulate information about the pod count and container port name (values), in addition to prefix and exact name sets (keys) for the `WorkloadSelector`.